pkg/kubernetes/kube.go

```diff
@@ -0,0 +1,438 @@
+	var (
+		home, _        = os.UserHomeDir()
+		kubeconfigPath = filepath.Join(home, ".kube", "config")
+	)
+
+	if exported := os.Getenv("KUBECONFIG"); exported != "" {
+		kubeconfigPath = exported
+	}
+
+	if _, err := os.Stat(kubeconfigPath); os.IsNotExist(err) {
+		return errors.Errorf("BDD >> expected kubeconfig to exist for create operation, '%v'", kubeconfigPath)
+	}
+
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfigPath)
+	if err != nil {
+		return err
+	}
+
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	dynClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		log.Fatal("Unable to construct dynamic client", err)
+	}
+
+	_, err = client.Discovery().ServerVersion()
+	if err != nil {
+		return err
+	}
+
+	kc.KubeInterface = client
+	kc.DynamicInterface = dynClient
+	kc.DiscoveryInterface = discoveryClient
+	//kc.RESTConfig = config
+
+	return nil
+}
+
+// Operations supported: create, submit, delete
+func (kc *Client) ResourceOperation(operation, resourceFileName string) error {
+
+	resourcePath := filepath.Join("templates", resourceFileName)
+	args := util.NewTemplateArguments()
+
+	gvr, resource, err := util.GetResourceFromYaml(resourcePath, kc.DiscoveryInterface, args)
+	if err != nil {
+		return err
+	}
+
+	switch operation {
+	case OperationCreate, OperationSubmit:
+		_, err = kc.DynamicInterface.Resource(gvr.Resource).Namespace(resource.GetNamespace()).Create(resource, metav1.CreateOptions{})
+		if err != nil {
+			if kerrors.IsAlreadyExists(err) {
+				// already created
+				break
+			}
+			return err
+		}
+	case OperationDelete:
+		err = kc.DynamicInterface.Resource(gvr.Resource).Namespace(resource.GetNamespace()).Delete(resource.GetName(), &metav1.DeleteOptions{})
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				// already deleted
+				break
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+// States supported: created and deleted
+func (kc *Client) ResourceShouldBe(resourceFileName, state string) error {
+	var (
+		exists  bool
+		counter int
+	)
+	resourcePath := filepath.Join("templates", resourceFileName)
+	args := util.NewTemplateArguments()
+	gvr, resource, err := util.GetResourceFromYaml(resourcePath, kc.DiscoveryInterface, args)
+	if err != nil {
+		return err
+	}
+
+	for {
+		exists = true
+		if counter >= DefaultWaiterRetries {
+			return errors.New("waiter timed out waiting for resource state")
+		}
+		log.Infof("BDD >> waiting for resource %v/%v to become %v", resource.GetNamespace(), resource.GetName(), state)
+
+		_, err := kc.DynamicInterface.Resource(gvr.Resource).Namespace(resource.GetNamespace()).Get(resource.GetName(), metav1.GetOptions{})
+		if err != nil {
+			if !kerrors.IsNotFound(err) {
+				return err
+			}
+			log.Infof("BDD >> %v/%v is not found: %v", resource.GetNamespace(), resource.GetName(), err)
```

Comment:
Can we change anywhere it starts with `BDD >>` to `[KUBEDOG]`, WDYT? might make more sense for logging

Reply:
I think its a great idea